sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java

```diff
@@ -217,26 +227,31 @@ protected AvroCoder(Class<T> type, Schema schema) {
     this.decoder = new EmptyOnDeserializationThreadLocal<>();
     this.encoder = new EmptyOnDeserializationThreadLocal<>();
 
-    // Reader and writer are allocated once per thread and are "final" for thread-local Coder
-    // instance.
-    this.reader = new EmptyOnDeserializationThreadLocal<DatumReader<T>>() {
-      private final AvroCoder<T> myCoder = AvroCoder.this;
-      @Override
-      public DatumReader<T> initialValue() {
-        return myCoder.getType().equals(GenericRecord.class)
-            ? new GenericDatumReader<T>(myCoder.getSchema())
-            : new ReflectDatumReader<T>(myCoder.getSchema());
-      }
-    };
-    this.writer = new EmptyOnDeserializationThreadLocal<DatumWriter<T>>() {
-      private final AvroCoder<T> myCoder = AvroCoder.this;
-      @Override
-      public DatumWriter<T> initialValue() {
-        return myCoder.getType().equals(GenericRecord.class)
-            ? new GenericDatumWriter<T>(myCoder.getSchema())
-            : new ReflectDatumWriter<T>(myCoder.getSchema());
-      }
-    };
+    // Reader and writer are allocated once per thread per Coder
+    this.reader =
+        new EmptyOnDeserializationThreadLocal<DatumReader<T>>() {
+          private final AvroCoder<T> myCoder = AvroCoder.this;
+
+          @Override
+          public DatumReader<T> initialValue() {
+            return myCoder.getType().equals(GenericRecord.class)
+                ? new GenericDatumReader<T>(myCoder.getSchema())
+                : new ReflectDatumReader<T>(
+                    myCoder.getSchema(), myCoder.getSchema(), myCoder.getReflectData());
```

Comment:
I think the fix works because internally the readers where using `SpecificData.getClass()` or some similar code somewhere and this was using the class loader that was cached in the `ReflectData` singleton. Now we have a new `ReflectData` with the proper class loader.